### PR TITLE
use modern JavaScript features

### PR DESCRIPTION
**Pull request by staltz**
Just a bunch of these:

- Replace `var` with `let` or `const`
- Use arrow functions where convenient (and where we don't use `this`)
- Use array methods instead of creating our own
- Use `??` instead of `||`
- Remove `isFunction`, `isString`, etc

Nothing really TypeScript here yet.

### CODE REVIEW

**Comment by arj03** [comment]
Looooooooove this! Reviewed, looks good. I'm just reading the code here, I'm assuming you have run the tests :)

**Comment by arj03** [comment]
Net: -45 lines, 45 bonus points for @staltz

**Comment by staltz** [comment]
Yes, I have run the tests. (By the way we should enable CI if possible)

Thanks @arj03

**Comment by mixmix** [comment]
I didn't know about `??`, cool. 
Reading on MDN though it doesn't look like great support yet : 

image

specifically looking at that node 14 .... I'm on node 10 for most proejcts at the moment and gearing to jump to 12 once I know all my electron installers run smoothly ...

File: src/api.ts [hunk 1]
```diff
@@ -1,68 +1,56 @@
...
-function id<T = any> (e: T): T {
-  return e
-}
+const identity = (x: any): any => x
 
 function merge (a: any, b: any, mapper?: any) {
-  mapper = mapper || id
-
-  for (var k in b) {
+  mapper = mapper ?? identity
```
Comment (mixmix) [inline src/api.ts:8]: yeah just tested it, this breaks in node 10 + 12, which are both LTS versions I would expect to work.

I don't know what the output of this .ts code is though ...

Comment (staltz) [reply src/api.ts:8]: @mixmix No worries, TypeScript compiles down to ES5, so we're looking good in terms of support. I considered configuring TypeScript to compile to ES6, which would support node 10 and above, and *partially support* node 5, but I think it's important to keep the support broad when it comes to secret-stack.

Here's how `src/api.ts` compiles down to `lib/api.js` (first lines where it's relevant to your comment):

```js
"use strict";
var u = require("./util");
var EventEmitter = require('events');
var Hookable = require('hoox');
function id(e) {
    return e;
}
function merge(a, b, mapper) {
    mapper = mapper || id;
    for (var k in b) {
        if (b[k] &&
            typeof b[k] === 'object' &&
            !Buffer.isBuffer(b[k]) &&
            !Array.isArray(b[k])) {
            merge((a[k] = {}), b[k], mapper);
        }
        else {
            a[k] = mapper(b[k], k);
        }
    }
    return a;
}

...
```

Comment (mixmix) [reply src/api.ts:8]: Oh that's awesome news

**Comment by staltz** [comment]
Thanks @arj03 and @mixmix for reviewing :heart:

**Pull request merged**